comp/core/flare/builder/builder.go

```diff
@@ -147,19 +152,55 @@ type FlareBuilder interface {
 	Save() (string, error)
 }
 
-// FlareCallback is a function that can be registered as a data provider for flares. This function, if registered, will
-// be called everytime a flare is created.
+// FlareArgs contains the arguments passed in to a specific flare generation request.
+// All consumers of FlareArgs should be able to safely ingest an empty (default) struct.
+type FlareArgs struct {
+	StreamLogsDuration   time.Duration // Add stream-logs data to the flare. It will collect logs for the amount of seconds passed to the flag
+	ProfileDuration      time.Duration // Add performance profiling data to the flare. It will collect a heap profile and a CPU profile for the amount of seconds passed to the flag, with a minimum of 30s
+	ProfileMutex         bool          // Add mutex profile to the performance data in the flare
+	ProfileMutexFraction int           // Set the fraction of mutex contention events that are reported in the mutex profile
+	ProfileBlocking      bool          // Add gorouting blocking profile to the performance data in the flare
+	ProfileBlockingRate  int           // Set the fraction of goroutine blocking events that are reported in the blocking profile
+}
+
+// FlareCallback is a function that can be registered as a data provider for flares by way of the FlareProvider struct.
+// This function, if registered, will be called everytime a flare is created.
 type FlareCallback func(fb FlareBuilder) error
 
+// FlareTimeout is a function that provides the maximum expected runtime duration of a FlareProvider's callback.
+// Return 0 from this function to utilize the default timeout instead.
+type FlareTimeout func(fb FlareBuilder) time.Duration
+
+// FlareFiller is a struct that can be registered as a data provider for flares. This struct's callback, if registered, will
+// be called everytime a flare is created.
+type FlareFiller struct {
+	Callback FlareCallback
+	Timeout  FlareTimeout
+}
```

Comment:
Why do we need a new type `FlareFiller` ? Could we not extend the `Provider` type to include a timeout ?

Also, why is `Timeout` a function that return a `Duration` rather than a `Duration` directly ? Using a Duration would mean we could also rely on it's default value and remove the `DefautlFiller` function.

Reply:
The new type is to firmly link the callbacks with their timeout. My understanding of fx groups is that the "Provider" struct is effectively lost when the Callbacks are all aggregated in to the dependencies.Providers struct value of [flare.go](https://github.com/DataDog/datadog-agent/blob/main/comp/core/flare/flare.go#L51). If there's a way to keep the association of specific timeout + callback without resorting to a new type I would be rather intensely interested in hearing more. 

As for function vs duration, the issue I run in to with a raw duration is that the current implementations of both the streamlogs and the profiling features offer a user customizable timeout. While we might be able to work in a reasonable "upper bound" duration, I do think this implementation makes timeout issues harder to miss.